ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/Genesis/Setup/Classifiers.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Test.Ouroboros.Consensus.Genesis.Setup.Classifiers (
+    Classifiers (..)
+  , classifiers
+  ) where
+
+import           Ouroboros.Consensus.Block.Abstract (SlotNo (SlotNo),
+                     withOrigin)
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Network.AnchoredFragment (anchor, anchorToSlotNo,
+                     headSlot)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.AnchoredFragment.Extras (slotLength)
+import           Test.Ouroboros.Consensus.BlockTree (BlockTree (..),
+                     BlockTreeBranch (..))
+import           Test.Ouroboros.Consensus.PointSchedule
+import           Test.Util.Orphans.IOLike ()
+
+{- | Interesting classification predicates used by the test:
+
+1. the immutable tip should ALWAYS be honest
+2. the selection should be honest IF there is at least a Genesis window after the intersection.
+
+It is possible that, at the end of a test, the selection might be dishonest.
+Consider the following example, where k = 5 and scg = 12:
+
+    slots:  0  1  2  3  4  5  6  7  8  9  10  11  12
+    trunk:  0──1──2──3──4──5──6──7──8──────9──10──11
+                     ╰──4──5──6──7──8──9──────────10
+
+Because of the Limit on Eagerness, the selection is stuck at slot 8 (k-th
+block after the intersection).
+Since the test chains only have 7 and 8 blocks after the intersection, Genesis
+will not be able to make a final decision, since it needs to have enough information
+to determine that one of the fragments is definitely denser within 12 slots, and here
+both fragments have the potential to win the race.
+This means that a test will terminate while the selection contains an arbitrary fragment
+whose tip is the block in slot 8.
+
+Without Genesis, if the blocks are sent so that the N+1st slot's blocks arrive after all
+blocks in slot N have arrived, the adversarial fragment will be selected every time because
+it will have k+1 blocks first.
```

Comment:
I think it is unclear in which order the different blocks are sent.

Reply:
it took us like 15 minutes to write this sentence because we couldn't figure out how to express this clearly 😅